### ISSUE TITLE
Qubes Builder fails with permission error when SELinux is enabled

### ISSUE THREAD

**Issue by permezel** [QubesOS/qubes-issues#6376]
**Qubes OS version**

4.0

**Affected component(s) or functionality**

QubesBuilder

**Brief summary**

It fails with error:
```
-> Preparing fc32 build environment
-> Initializing RPM database...
error: can't create transaction lock on /home/dap/qubes-builder/chroot-dom0-fc32/var/lib/rpm/.rpm.lock (Permission denied)
```
This is because it is not running as root but ```chroot-dom0-fc32``` is owned by root, and the Makefile(s) do not attempt to use ```sudo```.  Also, there is nothing else in the chroot root.

```
~/qubes-builder $ ls -Rl chroot-dom0-fc32
chroot-dom0-fc32:
total 0
~/qubes-builder $ ls -ld chroot-dom0-fc32
drwxr-xr-x   1 root           root            0 2021-01-30 17:21 chroot-dom0-fc32
```

**How Reproducible**

100%

**To Reproduce**

1. Follow the steps in [here](https://www.qubes-os.org/doc/qubes-builder/)
2. 
3. 

**Expected behavior**

Clean build.

**Actual behavior**

```make cubes``` fails.

**Screenshots**

[make.out.txt](https://github.com/QubesOS/qubes-issues/files/5897555/make.out.txt)
[builder.conf.txt](https://github.com/QubesOS/qubes-issues/files/5897558/builder.conf.txt)

**Additional context**

**Solutions you've tried**

I've tried running setup.
I've tried setting the prefix to marmarek.  That fails with a different error.
```-> Updating sources for linux-kernel...
--> Fetching from https://github.com/marmarek/qubes-linux-kernel.git stable-5.4...
fatal: Remote branch stable-5.4 not found in upstream origin
make: *** [Makefile:213: linux-kernel.get-sources] Error 1
```
If I ignore the above error and run ```make qubes``` it also fails with the chroot-fc32-dom0 problem.

**Relevant [documentation](https://www.qubes-os.org/doc/) you've consulted**

**Related, [non-duplicate](https://www.qubes-os.org/doc/reporting-bugs/#new-issues-should-not-be-duplicates-of-existing-issues) issues**

**Comment by unman** [comment]
All you say is true, but it works in Qubes. The conclusion has to be
that you have not created the build environment correctly.
*That* would be useful information in the "additional context" section.

Incidentally, there's no such thing as "cubes" here - the project is
Qubes, and the command would be `make qubes`

**Comment by permezel** [comment]
"it works in Qubes" is not particularly helpful.  Per the docs, any rpm-based distro is supposed to work.
I just spun up a Fedora 32 VM under the Fed33 system in ```Boxes``` and carefully followed the steps in the https://www.qubes-os.org/doc/qubes-iso-building/ document.  In that case, ```make get-sources``` fails repeatedly.

```
--> Downloading additional sources for xscreensaver...
make[1]: *** [Makefile:12: xscreensaver-5.43.tar.gz] Error 8
make: *** [Makefile:215: xscreensaver.get-sources-extra] Error 2
```

If I ignore that error, and run ```make qubes``` it again fails as before:
```
make[1]: Entering directory '/home/dap/qubes-builder'
-> Preparing fc25 build environment
-> Initializing RPM database...
error: can't create transaction lock on /home/dap/qubes-builder/chroot-dom0-fc25/var/lib/rpm/.rpm.lock (Permission denied)
make[1]: *** [/home/dap/qubes-builder/qubes-src/builder-rpm/Makefile-legacy.rpmbuilder:44: /home/dap/qubes-builder/chroot-dom0-fc25/home/user/.prepared_base] Error 1
make[1]: Leaving directory '/home/dap/qubes-builder'
make: *** [Makefile:262: gcc-dom0] Error 1
```

Regarding spelling errors: I reserve the wright to make as meny spelling errs as I might happen to be inclided or desposed to do.  But I do attempt to prufe read an corrext.  Cirrioussly.

I was hoping to use the NUC10 victim for building Qubes and testing to potentially assist in resolving the 4.0 and 4.1 initial boot failures.  I do have an older NUC7i5BNH system I am running Qubes on.  My next step will be to spin up a Fedora VM under qubes and see how far I get there trying to build.

Here is some "additional context":
To the best of my knowledge, I *have indeed* created the build environment correctly.
If the conclusion "has to be that I created it incorrectly" then where are the correct documents which describe the correct way to create it?

As fas as the xscreensaver failure is concerned, this is due to the version having been updated.  This is the command which is being issued and which fails:
```
wget --no-use-server-timestamps -q -O xscreensaver-5.43.tar.gz.UNTRUSTED https://www.jwz.org/xscreensaver/xscreensaver-5.43.tar.gz
```

If you go onto that site, it appears that 5.45 is the most recent version.  5.44 is available using ```wget``` and 5.45, but not 5.43.

This is clearly *not* my build environment, I respectfully submit.

**Comment by marmarek** [comment]
> "it works in Qubes" is not particularly helpful. Per the docs, any rpm-based distro is supposed to work.

Yes, it should work just fine on any rpm-based distro. In fact should also work on Debian too, but we test that much less. Thanks for the report!

> make[1]: Entering directory '/home/dap/qubes-builder'
> -> Preparing fc25 build environment
> -> Initializing RPM database...
> error: can't create transaction lock on /home/dap/qubes-builder/chroot-dom0-fc25/var/lib/rpm/.rpm.lock (Permission denied)

Internally, the script that calls it, uses sudo, so it should have access... Do you have some unusual sudo configuration (like non-root default user or such)? Another possibility is that SELinux (enabled by default on Fedora) gets in the way - you can check that by looking into `/var/log/audit/audit.log`, or by temporarily disable it with `setenforce 0`.

> If you go onto that site, it appears that 5.45 is the most recent version. 5.44 is available using `wget` and 5.45, but not 5.43.

Thanks, indeed seems the upstream blocks downloading anything but the few recent versions. I've adjusted makefile to use a mirror.

**Comment by permezel** [comment]
Changing the URL line in the Makefile works.
From
```
URL := https://www.jwz.org/xscreensaver/xscreensaver-$(VERSION).tar.gz
```
to
```
URL := https://ftp.osuosl.org/pub/blfs/conglomeration/xscreensaver/xscreensaver-$(VERSION).tar.gz
```

**Comment by permezel** [comment]
> Internally, the script that calls it, uses sudo, so it should have access... Do you have some unusual sudo configuration (like non-root default user or such)? Another possibility is that SELinux (enabled by default on Fedora) gets in the way - you can check that by looking into `/var/log/audit/audit.log`, or by temporarily disable it with `setenforce 0`.
>
I do have ```sudo``` setup so that ```wheel``` (admin) group members do not need a password.
I checked the audit log, but it was APL to me.  The `setenforce 0` did the trick, though.

I suggest this be added to the docs, and/or to the Makefile.  The first command which is failing should detect that it failed, and then possibly suggest disabling SELinux.

**Comment by permezel** [comment]
`setenforce 0` allows me to proceed on both the `Fedora 32 Boxes VM` and the `Fedora 33 on NUC10` builds.

I maintain that either the docs or the Makefiles or both should catch this.  I found someone ran into the same error as me 8 days prior to me having done so, per the `goog`.

**Comment by DemiMarie** [comment]
SELinux needing to be disabled is itself a bug (#4239).  What denials do you get in the audit log?

**Comment by permezel** [comment]
> SELinux needing to be disabled is itself a bug (#4239). What denials do you get in the audit log?

That bug pertains to a running Qubes system.  This bug pertains to SELinux silently impacting builds in non-Qubes systems.
The commands which fail have something to do with chrooted rpm builds failng to construct the necessary chrooted filesystem.  Subsequent commands fail because some path does not exist.  It would be helpful if the initial chrooted command either detected the failure and caused the `make qubes` to fail at that point, or if at that build step there was a verification step that the expected filesystem was created and a build failure with perhaps a suggestion to disable SELinux.

**Comment by iprid** [comment]
Perhaps this ticket should be renamed to - SELinux prevents Qubes Builder from working. I mean, this ticket is not related to building an ISO so much as to building the packages themselves.

**Comment by DemiMarie** [comment]
> > SELinux needing to be disabled is itself a bug (#4239). What denials do you get in the audit log?
> 
> That bug pertains to a running Qubes system. This bug pertains to SELinux silently impacting builds in non-Qubes systems.
> The commands which fail have something to do with chrooted rpm builds failng to construct the necessary chrooted filesystem. Subsequent commands fail because some path does not exist. It would be helpful if the initial chrooted command either detected the failure and caused the `make qubes` to fail at that point, or if at that build step there was a verification step that the expected filesystem was created and a build failure with perhaps a suggestion to disable SELinux.

I still consider having to disable SELinux to be a bug.  Once #4239 is fixed, SELinux will be enabled by default.

**Comment by andrewdavidwong** [comment]
> Perhaps this ticket should be renamed to - SELinux prevents Qubes Builder from working. I mean, this ticket is not related to building an ISO so much as to building the packages themselves.

Updated to a variant of your suggestion.

**Comment by iprid** [comment]
@andrewdavidwong thanks, I see that there are other SELinux related bugs, so perhaps this will help us group them. And also improve context.

**Comment by 51lieal** [comment]
@DemiMarie, Using fedora 35 (not qubes) and selinux-policy-35.3-1.20211019git94970fc.fc35.noarch

My log is here, hope this helps. : 
----
time->Sun Dec 12 03:06:15 2021
type=AVC msg=audit(1639253175.829:452): avc:  denied  { dac_read_search } for  pid=4169 comm="rpmdb" capability=2  scontext=unconfined_u:unconfined_r:rpmdb_t:s0-s0:c0.c1023 tcontext=unconfined_u:unconfined_r:rpmdb_t:s0-s0:c0.c1023 tclass=capability permissive=0
----
time->Sun Dec 12 03:06:15 2021
type=AVC msg=audit(1639253175.829:453): avc:  denied  { dac_override } for  pid=4169 comm="rpmdb" capability=1  scontext=unconfined_u:unconfined_r:rpmdb_t:s0-s0:c0.c1023 tcontext=unconfined_u:unconfined_r:rpmdb_t:s0-s0:c0.c1023 tclass=capability permissive=0
----
time->Sun Dec 12 03:06:15 2021
type=AVC msg=audit(1639253175.829:454): avc:  denied  { dac_read_search } for  pid=4169 comm="rpmdb" capability=2  scontext=unconfined_u:unconfined_r:rpmdb_t:s0-s0:c0.c1023 tcontext=unconfined_u:unconfined_r:rpmdb_t:s0-s0:c0.c1023 tclass=capability permissive=0
----
time->Sun Dec 12 03:06:15 2021
type=AVC msg=audit(1639253175.829:455): avc:  denied  { dac_override } for  pid=4169 comm="rpmdb" capability=1  scontext=unconfined_u:unconfined_r:rpmdb_t:s0-s0:c0.c1023 tcontext=unconfined_u:unconfined_r:rpmdb_t:s0-s0:c0.c1023 tclass=capability permissive=0

**Comment by lankredotp41** [comment]
This issue still affects Qubes OS 4.2 with fedora-38. I needed to set `setenforce 0` for build to succeed.

**Comment by marmarek** [comment]
SELinux compatibility is not planned for legacy builder (https://github.com/QubesOS/qubes-builder). It is planned for builder v2 (https://github.com/QubesOS/qubes-builderv2), but also doesn't work yet (https://github.com/QubesOS/qubes-issues/issues/8547)

**Comment by andrewdavidwong** [comment]
This issue has been closed as "won't fix." If anyone has a new reason for why this should be done, please leave a comment, and we'll be happy to take another look. Thank you.

**Comment by lankredotp41** [comment]
> SELinux compatibility is not planned for legacy builder (https://github.com/QubesOS/qubes-builder) - `setenforce 0` is necessary there.

In that case this should be mentioned in the legacy builder README so others won't search all around for why it fails.

**Comment by ben-grande** [comment]
I don't think this issue should remain open, no change to V1 code or docs will be done. About dropping the V1 docs, it will be hidden or removed in:

- https://github.com/QubesOS/qubes-issues/issues/9930
